hibernate-core/src/main/java/org/hibernate/boot/spi/BasicTypeRegistration.java

```diff
@@ -25,7 +29,23 @@ public BasicTypeRegistration(BasicType basicType) {
 
 	public BasicTypeRegistration(BasicType basicType, String[] registrationKeys) {
 		this.basicType = basicType;
-		this.registrationKeys = registrationKeys;
+		Class returnedClass = basicType.getReturnedClass();
+
+		if ( returnedClass != null ) {
+			String returnedClassName = returnedClass.getName();
+
+			List<String> registrationKeysList = registrationKeys == null || registrationKeys.length == 0 ?
+					new ArrayList<>() :
+					new ArrayList<>( Arrays.asList( registrationKeys ) );
+
+			if ( !registrationKeysList.contains( returnedClassName ) ) {
+				registrationKeysList.add( returnedClassName );
+			}
+			this.registrationKeys = registrationKeysList.toArray( new String[] {} );
```

Comment:
There might be some minor performance optimization opportunity here when class name can be found in the array. For instance
```
boolean found = // maybe by some array helper method
if ( !found) {
    // go about clumsy array appending, maybe by some array helper method as well
}
```
That way we can some unnecessary transforming between array and list when the class name has been found within the array.